File: python/acer/game.py
import datetime
import random

from acer.player import Player


class Game:
    def __init__(self):
        self.gameTime = datetime.datetime.now()
        self.numberOfPlayers = 0
        self.numberOfDays = 0
        self.players = []
        self.shuffledDays = []

    def setup(self):
        self.setup_game()
        self.setup_players()

    def get_number_of_days_per_player(self):
        return self.numberOfDays / self.numberOfPlayers

    def get_player(self, random_player_position):
        return next((x for x in self.players if x.position == random_player_position), None)

    def still_has_available_days_to_pick(self):
        total_days_picked = 0

        for p in self.players:
            total_days_picked += len(p.days_to_choose)

        return total_days_picked < self.numberOfDays

    def setup_game(self):
        self.numberOfPlayers = int(input("Number of Players? "))
        self.numberOfDays = int(input("Number of Days? "))

        shuffled_days = list(range(1, self.numberOfDays + 1))

        # shuffle them N times...
        for r in range(1, self.numberOfDays + 1):
            random.shuffle(shuffled_days)

        self.shuffledDays = shuffled_days
        print()

    def setup_players(self):
        players = []

        for i in range(1, self.numberOfPlayers + 1):
            player_name = input(f"Name of Player {i}? ")
            player = Player(i, player_name)

            players.append(player)

        self.players = players
        print()

    def choose_days_to_open(self):
        while self.still_has_available_days_to_pick():
            random_player_position = random.randrange(1, self.numberOfPlayers + 1)
            current_player = self.get_player(random_player_position)

            if len(current_player.days_to_choose) == self.get_number_of_days_per_player():
                continue

            input(f"{current_player.name}, hit <enter> to pick a number... ")

            day_to_open = self.shuffledDays[0]
            print(f"{current_player.name}, you got day {day_to_open}")
            current_player.days_to_choose.append(day_to_open)

            self.shuffledDays.pop(0)
            if len(self.shuffledDays) == 0:
                print("\n[All Days Picked!]\n")
            else:
                print(f"[Days remaining: {len(self.shuffledDays)}]\n")

File: golang/game.go
package main

import (
	"errors"
	"fmt"
	"log"
	"math/rand"
	"strconv"
	"time"
)

type Game struct {
	//gameTime        time.Time `default:"time.Now()"`
	NumberOfPlayers       int `default:"0"`
	NumberOfDays          int `default:"0"`
	NumberOfDaysPerPlayer []int
	Players               []*Player
	ShuffledDays          []int
}

func New(p Printer) Game {
	var players []*Player

	p.PrintNumberOfPlayersPrompt()
	var numberOfPlayers int
	_, _ = fmt.Scanln(&numberOfPlayers)

	p.PrintNumberOfDaysPrompt()
	var numberOfDays int
	_, _ = fmt.Scanln(&numberOfDays)

	var shuffledDays = make([]int, numberOfDays)
	rand.Seed(time.Now().UnixNano())
	for i := 0; i < numberOfDays; i++ {
		shuffledDays[i] = i + 1
	}

	for i := 0; i < numberOfDays; i++ {
		rand.Shuffle(len(shuffledDays), func(i, j int) { shuffledDays[i], shuffledDays[j] = shuffledDays[j], shuffledDays[i] })
	}

	numberOfDaysPerPlayer := distributeDays(numberOfDays, numberOfPlayers)

	for i := 1; i <= numberOfPlayers; i++ {
		var playerName string
		fmt.Printf("Name of Player %d? ", i)
		_, _ = fmt.Scanln(&playerName)

		var d = numberOfDaysPerPlayer[i-1]
		player := NewPlayer(i, playerName, d)

		players = append(players, player)
	}

	g := Game{numberOfPlayers, numberOfDays, numberOfDaysPerPlayer, players, shuffledDays}

	return g
}

func distributeDays(numberOfDays int, numberOfPlayers int) []int {
	d := numberOfDays / numberOfPlayers
	r := numberOfDays % numberOfPlayers
	var dist []int

	for i := 0; i < numberOfPlayers; i++ {
		dist = append(dist, d)
	}

	for j := 0; j < r; j++ {
		dist[j] = dist[j] + 1
	}

	return dist
}

func (g *Game) ChooseDaysToOpen(p Printer) {
	rand.Seed(time.Now().UnixNano())

	for ok := true; ok; ok = len(g.ShuffledDays) > 0 {
		//1. Pick random player, weighted by number of pics remaining (?)
		var randomPlayerPosition = rand.Intn(len(g.Players)-1+1) + 1

		currentPlayer, err := getPlayer(g.Players, randomPlayerPosition)
		if err != nil {
			log.Fatal(err)
		}

		if currentPlayer.allDaysPicked() {
			//fmt.Printf("Done picking for %s\n", currentPlayer.name)
			continue
		}

		//2. "Player X, hit <enter> to stop the spinner" (have spinning icon)
		p.printRequestPlayerInteraction(currentPlayer.name)
		_, _ = fmt.Scanln()

		//3. On <enter>, get the next day from the shuffled list of days.
		var dayToOpen = g.ShuffledDays[0]
		p.printPlayerTurnResult(currentPlayer.name, dayToOpen)

		//4. Assign random number to Player X.
		currentPlayer.DaysToChoose = append(currentPlayer.DaysToChoose, dayToOpen)

		//5. Remove day just assigned to the user.
		g.ShuffledDays = g.ShuffledDays[1:]
		p.printDaysRemaining(len(g.ShuffledDays))
	}
}

func getPlayer(players []*Player, randomPlayerPosition int) (*Player, error) {
	for _, p := range players {
		if p.position == randomPlayerPosition {
			return p, nil
		}
	}

	return &Player{}, errors.New("no player found at " + strconv.Itoa(randomPlayerPosition))
}
